File: 11 robots - market maker amend/python/market_maker_amend.py
import os
import sys

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "..", "helpers", "python"))
import tse_helpers as H

tse = H.tse

SYMBOL = "MMFLOW"
QUIET_GAP_NANOSECONDS = 10_000_000_000
PRINT_LAG_NANOSECONDS = 500_000_000
SAME_SIDE_RUN_LENGTH = 3
SWEEP_MULTIPLE = 3.0
QUOTE_SIDE = int(tse.Side.Long)


def load_client_flow():
    rows = []
    with open(H.data_path("mm_client_flow.csv")) as handle:
        handle.readline()
        for line in handle:
            parts = line.strip().split(",")
            if len(parts) < 8:
                continue
            side = tse.Side.Long if parts[3] == "bid" else tse.Side.Short
            rows.append((int(parts[0]), int(side), float(parts[6]), float(parts[7])))
    return rows


def main():
    quote_price = 43.22
    step_away_price = 43.21
    full_size = 4.0
    defensive_size = 2.0
    inventory_cap = 12.0

    account = H.account("MarketMakerAmendExample", tse.StorageRegime.Mem, tse.LogLevel.Off)
    flow_mkt = account.create_market("client flow", tse.MdType.Book)
    print_mkt = account.create_market("prints", tse.MdType.Trade)
    account.create_simulator("sim", H.simulator_options(), 64, -1)
    account.add_contract(SYMBOL, 1, tse.Instrument.Equity, tse.Underlying.Equity, tse.Venue.Undefined, 10000)

    flow = {
        "quantity_sum": 0.0,
        "trade_count": 0,
        "run_length": 0,
        "previous_ts": 0,
        "previous_side": int(tse.Side.Undefined),
        "hit_side": int(tse.Side.Undefined),
        "sweep_fired": False,
        "run_fired": False,
        "gap_fired": False,
    }
    quote = {
        "resting_client_order_id": 0,
        "next_client_order_id": 1,
        "cancel_count": 0,
        "replace_count": 0,
        "modify_count": 0,
    }

    # Every executed client trade is measured once, here: an outsized print, a third
    # trade in a row on one side and a long silence are the three things the maker reacts to.
    def client_flow(storage, contract_id, message):
        if message.kind != tse.BookMessageKind.Executed:
            return False
        average = 0.0 if flow["trade_count"] == 0 else flow["quantity_sum"] / flow["trade_count"]
        flow["sweep_fired"] = flow["trade_count"] != 0 and message.quantity > SWEEP_MULTIPLE * average
        flow["gap_fired"] = (
            flow["trade_count"] != 0
            and message.tsNanoseconds - flow["previous_ts"] > QUIET_GAP_NANOSECONDS
        )
        flow["run_length"] = flow["run_length"] + 1 if message.txnSide == flow["previous_side"] else 1
        flow["run_fired"] = flow["run_length"] == SAME_SIDE_RUN_LENGTH
        if flow["run_fired"]:
            flow["run_length"] = 0
        flow["hit_side"] = message.txnSide
        flow["previous_side"] = message.txnSide
        flow["previous_ts"] = message.tsNanoseconds
        flow["quantity_sum"] += message.quantity
        flow["trade_count"] += 1
        storage.push(message.tsNanoseconds, message.quantity)
        return True

    account.add_input_book("ClientFlow", 4, tse.Duration.Nanoseconds, client_flow, flow_mkt, [SYMBOL])

    # One quote lives at a time and the maker knows it by its client order id: a cancel drops
    # the id, a replace mints a fresh one, a modify keeps it. Zero means nothing rests.
    def quote_is_gone(input_label, ts_nanoseconds, value):
        if quote["resting_client_order_id"] != 0:
            return False
        quote["resting_client_order_id"] = quote["next_client_order_id"]
        quote["next_client_order_id"] += 1
        return True

    def sweep_hit(input_label, ts_nanoseconds, value):
        if (
            not flow["sweep_fired"]
            or flow["hit_side"] != QUOTE_SIDE
            or quote["resting_client_order_id"] == 0
        ):
            return False
        quote["resting_client_order_id"] = 0
        quote["cancel_count"] += 1
        return True

    def same_side_run(input_label, ts_nanoseconds, value):
        if not flow["run_fired"] or quote["resting_client_order_id"] == 0:
            return False
        quote["resting_client_order_id"] = quote["next_client_order_id"]
        quote["next_client_order_id"] += 1
        quote["replace_count"] += 1
        return True

    def quiet_gap(input_label, ts_nanoseconds, value):
        if not flow["gap_fired"] or quote["resting_client_order_id"] == 0:
            return False
        quote["modify_count"] += 1
        return True

    def inventory_full(input_label, ts_nanoseconds, value):
        state = account.get_position_state(SYMBOL)
        return state.side == int(tse.Side.Long) and state.quantity >= inventory_cap

    account.add_pattern_formula("QuoteIsGone", tse.Duration.Nanoseconds, ["ClientFlow"], quote_is_gone)
    account.add_pattern_formula("SweepHit", tse.Duration.Nanoseconds, ["ClientFlow"], sweep_hit)
    account.add_pattern_formula("SameSideRun", tse.Duration.Nanoseconds, ["ClientFlow"], same_side_run)
    account.add_pattern_formula("QuietGap", tse.Duration.Nanoseconds, ["ClientFlow"], quiet_gap)
    account.add_pattern_formula("InventoryFull", tse.Duration.Nanoseconds, ["ClientFlow"], inventory_full)

    # The three amend builders share one argument list but read different parts of it: cancel ignores
    # quantity and price, modify reads quantity only, replace reads both. Pull the quote when a sweep
    # hits its side, step it away and shrink it against a one-sided run, restore its size when quiet.
    account.add_rule_market(
        "RestQuote", tse.RuleType.Entry,
        tse.make_rule_params(
            tse.Quantity.Fixed, full_size,
            tse.Price.Limit, quote_price, 0.0, 0.0,
            QUOTE_SIDE, tse.Side.Neutral, tse.Tif.Day, 10,
        ),
        "QuoteIsGone", SYMBOL,
    )
    account.add_rule_cancel("PullQuote", SYMBOL, 0.0, 0.0, "SweepHit")
    account.add_rule_replace("StepAway", SYMBOL, defensive_size, step_away_price, "SameSideRun")
    account.add_rule_modify("BackToFullSize", SYMBOL, full_size, 0.0, "QuietGap")
    account.add_rule_market(
        "Unwind", tse.RuleType.Exit,
        tse.make_rule_params(
            tse.Quantity.All, 0.0,
            tse.Price.Market, 0.0, 0.0, 0.0,
            tse.Side.Short, tse.Side.Long, tse.Tif.Day, 10,
        ),
        "InventoryFull", SYMBOL,
    )

    account.add_robot("AmendingMaker", ["RestQuote", "PullQuote", "StepAway", "BackToFullSize", "Unwind"])
    account.portfolio_subscribe(print_mkt, SYMBOL)
    account.start("AmendingMaker")

    message_id = 0
    for ts_nanoseconds, side, price, quantity in load_client_flow():
        message_id += 1
        flow_mkt.push_book_by_name(
            SYMBOL,
            tse.make_book_message(
                tse.BookMessageKind.Executed, ts_nanoseconds,
                message_id, message_id,
                price, quantity, side,
            ),
        )
        print_mkt.push_trade_by_name(
            SYMBOL,
            tse.TseTickTrade(ts_nanoseconds + PRINT_LAG_NANOSECONDS, price, quantity, int(tse.Side.Trade)),
        )

    summary = account.get_summary()
    account.close()

    print("market maker amend: cancels={} replaces={} modifies={} netProfit={:.4f} trades={}".format(
        quote["cancel_count"], quote["replace_count"], quote["modify_count"],
        summary.totalNetProfit, summary.totalNumberOfTrades))
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: 11 robots - market maker amend/cpp/market_maker_amend.cpp
#include "tse_helpers.hpp"

#include <cstdint>
#include <cstdio>
#include <cstring>
#include <fstream>
#include <string>
#include <vector>

namespace {

	std::string const symbol {"MMFLOW"};

	std::int64_t constexpr
		quietGapNanoseconds {10000000000LL},
		printLagNanoseconds {500000000LL};

	std::size_t constexpr sameSideRunLength {3u};

	double constexpr sweepMultiple {3.0};

	tse::Side constexpr quoteSide {tse::Side::long_};

	struct ClientTrade final {
		std::int64_t tsNanoseconds;
		tse::Side    side;
		double       price;
		double       quantity;
	};

	struct FlowState final {
		double       quantitySum;
		std::size_t  tradeCount;
		std::size_t  runLength;
		std::int64_t previousTsNanoseconds;
		tse::Side    previousSide;
		tse::Side    hitSide;
		bool         sweepFired;
		bool         runFired;
		bool         gapFired;
	};

	struct QuoteState final {
		std::uint64_t restingClientOrderId;
		std::uint64_t nextClientOrderId;
		std::size_t   cancelCount;
		std::size_t   replaceCount;
		std::size_t   modifyCount;
	};

	std::vector<ClientTrade> loadClientFlow()
	{
		std::ifstream file {helpers::dataPath("mm_client_flow.csv")};
		std::vector<ClientTrade> out;
		std::string line;
		std::getline(file, line);
		while (std::getline(file, line)) {
			long long ts {0};
			char sideText[8] {};
			double
				price {0.0},
				quantity {0.0};
			if (std::sscanf(line.c_str(), "%lld,%*d,%*[^,],%7[^,],%*d,%*d,%lf,%lf", &ts, sideText, &price, &quantity) != 4) {
				continue;
			}
			out.push_back(ClientTrade
			{
				static_cast<std::int64_t>(ts),
				std::strcmp(sideText, "bid") == 0 ? tse::Side::long_ : tse::Side::short_,
				price,
				quantity
			});
		}
		return out;
	}

	tse::BookMessage executedMessage
	(
		ClientTrade const& trade,
		std::uint64_t messageId
	)
	{
		tse::BookMessage message;
		message.kind = tse::BookMessageKind::executed;
		message.tsNanoseconds = trade.tsNanoseconds;
		message.messageId = messageId;
		message.orderId = messageId;
		message.price = trade.price;
		message.quantity = trade.quantity;
		message.txnSide = trade.side;
		return message;
	}

}

int main()
{
	tse::setLogLevel(tse::LogLevel::none);

	double const
		quotePrice {43.22},
		stepAwayPrice {43.21},
		fullSize {4.0},
		defensiveSize {2.0},
		inventoryCap {12.0};

	tse::Account account {"MarketMakerAmendExample", tse::StorageRegime::mem};
	tse::Market const
		flowMkt {account.createMarket("client flow", tse::MdType::book)},
		printMkt {account.createMarket("prints", tse::MdType::trade)};
	account.createSimulator("sim", helpers::simulatorConfig(), 64, -1);
	account.addContract(symbol, 1, tse::Instrument::equity, tse::Underlying::equity, tse::Venue::undefined, 10000);

	FlowState flow {0.0, 0u, 0u, 0, tse::Side::undefined, tse::Side::undefined, false, false, false};
	QuoteState quote {0u, 1u, 0u, 0u, 0u};

	// Every executed client trade is measured once, here: an outsized print, a third
	// trade in a row on one side and a long silence are the three things the maker reacts to.
	account.addInputBook
	(
		"ClientFlow", 4, tse::Duration::nanoseconds,
		[&flow](tse::Storage const& storage, std::string const&, tse::BookMessage const& message) -> bool
		{
			if (message.kind != tse::BookMessageKind::executed) { return false; }
			double const average
				{flow.tradeCount == 0u ? 0.0 : flow.quantitySum / static_cast<double>(flow.tradeCount)};
			flow.sweepFired = flow.tradeCount != 0u and message.quantity > sweepMultiple * average;
			flow.gapFired = flow.tradeCount != 0u
				and message.tsNanoseconds - flow.previousTsNanoseconds > quietGapNanoseconds;
			flow.runLength = message.txnSide == flow.previousSide ? flow.runLength + 1u : 1u;
			flow.runFired = flow.runLength == sameSideRunLength;
			if (flow.runFired) { flow.runLength = 0u; }
			flow.hitSide = message.txnSide;
			flow.previousSide = message.txnSide;
			flow.previousTsNanoseconds = message.tsNanoseconds;
			flow.quantitySum += message.quantity;
			++flow.tradeCount;
			storage.push(message.tsNanoseconds, message.quantity);
			return true;
		},
		flowMkt, {symbol}
	);

	// One quote lives at a time and the maker knows it by its client order id: a cancel drops
	// the id, a replace mints a fresh one, a modify keeps it. Zero means nothing rests.
	account.addPatternFormula
	(
		"QuoteIsGone", tse::Duration::nanoseconds, {"ClientFlow"},
		[&quote](std::string const&, std::int64_t, double) -> bool
		{
			if (quote.restingClientOrderId != 0u) { return false; }
			quote.restingClientOrderId = quote.nextClientOrderId++;
			return true;
		}
	);
	account.addPatternFormula
	(
		"SweepHit", tse::Duration::nanoseconds, {"ClientFlow"},
		[&flow, &quote](std::string const&, std::int64_t, double) -> bool
		{
			if (not flow.sweepFired or flow.hitSide != quoteSide or quote.restingClientOrderId == 0u) { return false; }
			quote.restingClientOrderId = 0u;
			++quote.cancelCount;
			return true;
		}
	);
	account.addPatternFormula
	(
		"SameSideRun", tse::Duration::nanoseconds, {"ClientFlow"},
		[&flow, &quote](std::string const&, std::int64_t, double) -> bool
		{
			if (not flow.runFired or quote.restingClientOrderId == 0u) { return false; }
			quote.restingClientOrderId = quote.nextClientOrderId++;
			++quote.replaceCount;
			return true;
		}
	);
	account.addPatternFormula
	(
		"QuietGap", tse::Duration::nanoseconds, {"ClientFlow"},
		[&flow, &quote](std::string const&, std::int64_t, double) -> bool
		{
			if (not flow.gapFired or quote.restingClientOrderId == 0u) { return false; }
			++quote.modifyCount;
			return true;
		}
	);
	account.addPatternFormula
	(
		"InventoryFull", tse::Duration::nanoseconds, {"ClientFlow"},
		[&account, inventoryCap](std::string const&, std::int64_t, double) -> bool
		{
			tse::PositionState const state {account.getPositionState(symbol)};
			return state.side == tse::Side::long_ and state.quantity >= inventoryCap;
		}
	);

	// The three amend builders share one argument list but read different parts of it: cancel ignores
	// quantity and price, modify reads quantity only, replace reads both. Pull the quote when a sweep
	// hits its side, step it away and shrink it against a one-sided run, restore its size when quiet.
	account.addRuleMarket
	(
		"RestQuote", tse::RuleType::entry,
		tse::RuleParams
		{
			tse::QuantityMode::fixed, fullSize,
			tse::PriceType::limit, quotePrice, 0.0, 0.0,
			quoteSide, tse::Side::neutral, tse::Tif::day, 10
		},
		"QuoteIsGone", symbol
	);
	account.addRuleCancel("PullQuote", symbol, 0.0, 0.0, "SweepHit");
	account.addRuleReplace("StepAway", symbol, defensiveSize, stepAwayPrice, "SameSideRun");
	account.addRuleModify("BackToFullSize", symbol, fullSize, 0.0, "QuietGap");
	account.addRuleMarket
	(
		"Unwind", tse::RuleType::exit,
		tse::RuleParams
		{
			tse::QuantityMode::all, 0.0,
			tse::PriceType::market, 0.0, 0.0, 0.0,
			tse::Side::short_, tse::Side::long_, tse::Tif::day, 10
		},
		"InventoryFull", symbol
	);

	account.addRobot("AmendingMaker", {"RestQuote", "PullQuote", "StepAway", "BackToFullSize", "Unwind"});
	account.portfolioSubscribe(printMkt, symbol);
	account.start("AmendingMaker");

	std::vector<ClientTrade> const clientFlow {loadClientFlow()};
	std::uint64_t messageId {0};
	for (ClientTrade const& trade : clientFlow) {
		++messageId;
		flowMkt.pushBook(symbol, executedMessage(trade, messageId));
		printMkt.pushTrade
		(
			symbol,
			tse::TradeTick {trade.tsNanoseconds + printLagNanoseconds, trade.price, trade.quantity, tse::Side::trade}
		);
	}

	tse::Summary const summary {account.getSummary()};

	std::printf
	(
		"market maker amend: cancels=%zu replaces=%zu modifies=%zu netProfit=%.4f trades=%lld\n",
		quote.cancelCount,
		quote.replaceCount,
		quote.modifyCount,
		summary.totalNetProfit,
		static_cast<long long>(summary.totalNumberOfTrades)
	);
	return 0;
}
